In [63]:
import pandas as pd
import numpy as np

In [64]:
Titanic = pd.read_csv('Titanic2.csv')
Titanic.head(20)

Unnamed: 0  pclass  survived  \
0            0       1         1   
1            1       1         1   
2            2       1         0   
3            3       1         0   
4            4       1         0   
5            5       1         1   
6            6       1         1   
7            7       1         0   
8            8       1         1   
9            9       1         0   
10          10       1         0   
11          11       1         1   
12          12       1         1   
13          13       1         1   
14          14       1         1   
15          15       1         0   
16          16       1         0   
17          17       1         1   
18          18       1         1   
19          19       1         0   

                                                 name     sex      age  sibsp  \
0                       Allen, Miss. Elisabeth Walton  female  29.0000      0   
1                      Allison, Master. Hudson Trevor    male   0.9167      1   
2                        Allison, Miss. Helen Loraine  female   2.0000      1   
3                Allison, Mr. Hudson Joshua Creighton    male  30.0000      1   
4     Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000      1   
5                                 Anderson, Mr. Harry    male  48.0000      0   
6                   Andrews, Miss. Kornelia Theodosia  female  63.0000      1   
7                              Andrews, Mr. Thomas Jr    male  39.0000      0   
8       Appleton, Mrs. Edward Dale (Charlotte Lamson)  female  53.0000      2   
9                             Artagaveytia, Mr. Ramon    male  71.0000      0   
10                             Astor, Col. John Jacob    male  47.0000      1   
11  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female  18.0000      1   
12                      Aubart, Mme. Leontine Pauline  female  24.0000      0   
13                       Barber, Miss. Ellen 'Nellie'  female  26.0000      0   
14               Barkworth, Mr. Algernon Henry Wilson    male  80.0000      0   
15                                Baumann, Mr. John D    male      NaN      0   
16                           Baxter, Mr. Quigg Edmond    male  24.0000      0   
17    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0000      0   
18                              Bazzani, Miss. Albina  female  32.0000      0   
19                               Beattie, Mr. Thomson    male  36.0000      0   

    parch    ticket      fare    cabin embarked boat   body  \
0       0     24160  211.3375       B5        S    2    NaN   
1       2    113781  151.5500  C22 C26        S   11    NaN   
2       2    113781  151.5500  C22 C26        S  NaN    NaN   
3       2    113781  151.5500  C22 C26        S  NaN  135.0   
4       2    113781  151.5500  C22 C26        S  NaN    NaN   
5       0     19952   26.5500      E12        S    3    NaN   
6       0     13502   77.9583       D7        S   10    NaN   
7       0    112050    0.0000      A36        S  NaN    NaN   
8       0     11769   51.4792     C101        S    D    NaN   
9       0  PC 17609   49.5042      NaN        C  NaN   22.0   
10      0  PC 17757  227.5250  C62 C64        C  NaN  124.0   
11      0  PC 17757  227.5250  C62 C64        C    4    NaN   
12      0  PC 17477   69.3000      B35        C    9    NaN   
13      0     19877   78.8500      NaN        S    6    NaN   
14      0     27042   30.0000      A23        S    B    NaN   
15      0  PC 17318   25.9250      NaN        S  NaN    NaN   
16      1  PC 17558  247.5208  B58 B60        C  NaN    NaN   
17      1  PC 17558  247.5208  B58 B60        C    6    NaN   
18      0     11813   76.2917      D15        C    8    NaN   
19      0     13050   75.2417       C6        C    A    NaN   

                          home.dest  ageNull  
0                      St Louis, MO        0  
1   Montreal, PQ / Chesterville, ON        0  
2   Montreal, PQ / Chesterville, ON        0  
3   Montreal, PQ / Chesterville, 

### 1. Sprawdź liczebność poszczególnych etykiet dla danych zmiennych jakościowych


In [65]:
#for col in Titanic:
    #print('Liczba etykiet zmiennej {}: {}'.format(col,len(Titanic[col].unique())))

print('Liczba etykiet zmiennej name: {}'.format(len(Titanic.name.unique())))
print('Liczba etykiet zmiennej pclass: {}'.format(len(Titanic.pclass.unique())))
print('Liczba etykiet zmiennej sibsp: {}'.format(len(Titanic.sibsp.unique())))
print('Liczba etykiet zmiennej sex: {}'.format(len(Titanic.sex.unique())))
print('Liczba etykiet zmiennej parch: {}'.format(len(Titanic.parch.unique())))
print('Liczba etykiet zmiennej embarked: {}'.format(len(Titanic.embarked.unique())))
print('Liczba etykiet zmiennej cabin: {}'.format(len(Titanic.cabin.unique())))
print('Liczba etykiet zmiennej boat: {}'.format(len(Titanic.boat.unique())))
print('Liczba etykiet zmiennej home.dest: {}'.format(len(Titanic['home.dest'].unique())))

Liczba etykiet zmiennej name: 1307
Liczba etykiet zmiennej pclass: 3
Liczba etykiet zmiennej sibsp: 7
Liczba etykiet zmiennej sex: 2
Liczba etykiet zmiennej parch: 8
Liczba etykiet zmiennej embarked: 4
Liczba etykiet zmiennej cabin: 187
Liczba etykiet zmiennej boat: 28
Liczba etykiet zmiennej home.dest: 370


### 2. Wyświetl z użyciem funkcji print liczbę wszystkich pasażerów.
### Wykorzystaj podobny sposób jak w punkcie 1.



In [66]:
print('Liczba pasażerów: {}'.format(len(Titanic)))

Liczba pasażerów: 1309


### 3. Skomentuj wyniki otrzymane w punkcie 1 i 2. Podziel zmienne ze
### względu na dużą i małą moc zbioru (kardynalność).

Analizując powyższe, można stwierdzić, że największą liczbność etykiet ma zmienna name. Niektóre zmienne tak jak płeć mają tylko 2 etykiety. Liczba pasażerów to 1309 osób.
Zmienne z dużą mocą zbioru: name, cabin, home.dest.
Zmienne z małą mocą zbioru: pclass, sibsp, sex, parche, embarked, boat.

### 4. Sprawdź, ile unikalnych etykiet ma zmienna mówiąca o kabinie danego
### pasażera. Użyj takiej funkcji, która zwraca wynik w postaci NumPy array.

In [67]:
print('Liczba unikalnych etykiet zmiennej mówiącej o kabinie danego pasażera: {}'.format(len(Titanic['cabin'].unique())))

Liczba unikalnych etykiet zmiennej mówiącej o kabinie danego pasażera: 187


In [68]:
print(type(Titanic['cabin'].unique()))

<class 'numpy.ndarray'>


In [69]:
Titanic['cabin'].unique()

array(['B5', 'C22 C26', 'E12', 'D7', 'A36', 'C101', nan, 'C62 C64', 'B35',
       'A23', 'B58 B60', 'D15', 'C6', 'D35', 'C148', 'C97', 'B49', 'C99',
       'C52', 'T', 'A31', 'C7', 'C103', 'D22', 'E33', 'A21', 'B10', 'B4',
       'E40', 'B38', 'E24', 'B51 B53 B55', 'B96 B98', 'C46', 'E31', 'E8',
       'B61', 'B77', 'A9', 'C89', 'A14', 'E58', 'E49', 'E52', 'E45',
       'B22', 'B26', 'C85', 'E17', 'B71', 'B20', 'A34', 'C86', 'A16',
       'A20', 'A18', 'C54', 'C45', 'D20', 'A29', 'C95', 'E25', 'C111',
       'C23 C25 C27', 'E36', 'D34', 'D40', 'B39', 'B41', 'B102', 'C123',
       'E63', 'C130', 'B86', 'C92', 'A5', 'C51', 'B42', 'C91', 'C125',
       'D10 D12', 'B82 B84', 'E50', 'D33', 'C83', 'B94', 'D49', 'D45',
       'B69', 'B11', 'E46', 'C39', 'B18', 'D11', 'C93', 'B28', 'C49',
       'B52 B54 B56', 'E60', 'C132', 'B37', 'D21', 'D19', 'C124', 'D17',
       'B101', 'D28', 'D6', 'D9', 'B80', 'C106', 'B79', 'C47', 'D30',
       'C90', 'E38', 'C78', 'C30', 'C118', 'D36', 'D48', 'D47', '

### 5. Zredukuj liczbę cech dla zmiennej opisującej kabiny poprzez zastąpienie obecnych etykiet w formacie PP11 do etykiet zawierających tylko pierwszą literę. Użyj astype(str).str[pozycja]. Nową zmienną nazwij CabinReduced. Wyświetl pierwsze 25 wierszy zbioru danych dla kolumn Cabin i CabinReduced


In [70]:
Titanic['CabinReduced'] = Titanic['cabin'].astype(str).str[0]

In [71]:
Titanic[['cabin','CabinReduced']].head(20)

cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9       NaN            n
10  C62 C64            C
11  C62 C64            C
12      B35            B
13      NaN            n
14      A23            A
15      NaN            n
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C

### 6. Wyświetl (jak w pkt 1) liczbę etykiet dla zmiennych z punkt 5. O ile
### procent zredukowano kardynalność zbioru zmiennej opisującej kabiny?

In [72]:
print('Liczba unikalnych etykiet zmiennej mówiącej o kabinie danego pasażera: {}'.format(len(Titanic['CabinReduced'].unique())))
reduced=(1-len(Titanic.CabinReduced.unique())/len(Titanic.cabin.unique()))*100
print('Kardynalność zbioru zmiennej opisującej kabiny zredukowano o: {} %'.format(round(reduced,2)))

Liczba unikalnych etykiet zmiennej mówiącej o kabinie danego pasażera: 9
Kardynalność zbioru zmiennej opisującej kabiny zredukowano o: 95.19 %


### 7. Uzasadnij, dlaczego dokonujesz redukcji akurat tej zmiennej. Jak to
### wpływa na przyszłe analizy. Czy powoduje jakieś negatywne skutki?

Dokonuję redukcji tej zmiennej, ponieważ to znacząco ułatwi zadanie modelu ML. Analiza będzie szybsza i łatwiejsza, ponieważ model będzie miał tylko 9 etykiet, a nie jak przed redukcją - 187. Negatywnym skutkiem takiej redukcji może być utrata części informacji o dokładnym położeniu kabiny, ponieważ po redukcji otrzymujemy informację tylko o sektorze/klasie kabiny.

In [73]:
Titanic.to_csv('Titanic3.csv')